### DISCUSSION TITLE
Retrieve AAL from OIDC provider

### DISCUSSION THREAD

**Discussion by alefray** [ory/kratos#2062] [Q&A] [answered]
Hi everyone,

We are using Kratos as an identity broker:
- OIDC provider (e.g. AzureAD) for employees or organizations
- email-password for users without entreprise-grade OIDC.

We have the requirement to enforce 2FA, i.e. have `aal2` level.
However, some users already have 2FA (e.g. TOTP) enabled directly at the OIDC provider level and some don't.

It feels not acceptable to require users to have two TOTP challenges : from the OIDC provider and from Kratos.

Is their any way to obtain an `aal2` level for exemple from the provider `acr` / `acm` claims ?

**Comment by vinckr** [comment]
Hello @alefray 
>It feels not acceptable to require users to have two TOTP challenges : from the OIDC provider and from Kratos.
I agree. 

Is there a way to skip the Kratos TOTP challenge if they have one from the OIDC provider? Or would that compromise security 🤔...

**Reply by alefray** [reply]
Hi, sorry for the late reply. 

Yes, skipping Kratos TOTP challenge would be fine.

In our scenario, Kratos acts as an identity federation between:
- Local users managed by Kratos - with password, TOTP etc.
- Federated users from multiple OIDC providers - without any password or TOTP managed by Kratos.

How can we make this distinction ?

How would it work when an application [request a second factor](https://www.ory.sh/docs/kratos/guides/two-factor-authentication-2fa-mfa#requesting-second-factor-authentication)?

**Reply by vinckr** [reply]
No problem.

I thought about it some more, and for example look at the Kratos session, for password for example: 

```
{
  "id": "171e5857-7532-4e17-a000-b8e50fac693a",
  "active": true,
  "expires_at": "2022-03-26T17:16:04.714191Z",
  "authenticated_at": "2022-03-23T17:16:04.764761Z",
  "authenticator_assurance_level": "aal1",
  "authentication_methods": [
    {
      "method": "password",
      "aal": "aal1",
      "completed_at": "2022-03-23T17:16:04.714287508Z"
    }
  ],
```
And for OIDC:
```
{
  "id": "c08c7ac5-f011-45c3-b51d-953a5fe7d092",
  "active": true,
  "expires_at": "2022-03-26T17:19:28.041555Z",
  "authenticated_at": "2022-03-23T17:19:28.041555Z",
  "authenticator_assurance_level": "aal1",
  "authentication_methods": [
    {
      "method": "oidc",
      "aal": "aal1",
      "completed_at": "2022-03-23T17:19:28.041549755Z"
    }
  ],
```


If I understand correctly, the challenge is that the `aal` in these session only really reflects the `aal` of Ory Kratos - for example in the second case the account in question was secured with a second factor.
I am not sure if this is something we can get from upstream providers, say how can Kratos know how the user authenticated at e.g. AzureAD?

`acm`/ `acr` claims are not standardized, so I guess this would need to be checked for each provider. For example for [Azure](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.web.claimconstants.acr?view=azure-dotnet), i am not sure what the acm claim returns.

Is say enforcing MFA (or skipping it) for the OIDC provider an option? 
And then you redirect all who don't sign up through the OIDC with "[request second factor](https://www.ory.sh/docs/kratos/guides/two-factor-authentication-2fa-mfa#requesting-second-factor-authentication)"
If you need to make the distinction on the OIDC provider (non-MFA / MFA users) it is harder, maybe [mutating with Oathkeeper](https://www.ory.sh/docs/oathkeeper/pipeline/mutator) and also depends what you get from the OIDC provider.

**Reply by aeneasr** [reply]
Most (basically almost all) OIDC providers (google,github,linkedin,facebook,….) do not support managing MFA for consuming apps, which is why we generally do not support MFA through social :/

**Comment by vinckr** [comment, marked as answer]
from the [above thread](https://github.com/ory/kratos/discussions/2062#discussioncomment-2424627):
>Most (basically almost all) OIDC providers (google,github,linkedin,facebook,….) do not support managing MFA for consuming apps, which is why we generally do not support MFA through social
